android/app/src/main/java/com/microsoft/codepush/react/CodePushPackage.java

```diff
@@ -178,6 +214,76 @@ public void downloadPackage(Context applicationContext, ReadableMap updatePackag
                 throw new CodePushUnknownException("Error closing IO resources.", e);
             }
         }
+
+        if (isZip) {
+            String unzippedFolderPath = getUnzippedFolderPath();
+            CodePushUtils.unzipFile(downloadFile, unzippedFolderPath);
+            CodePushUtils.deleteFileSilently(downloadFile);
+            String diffManifestFilePath = CodePushUtils.appendPathComponent(unzippedFolderPath,
+                    DIFF_MANIFEST_FILE_NAME);
+            File diffManifestFile = new File(unzippedFolderPath, DIFF_MANIFEST_FILE_NAME);
+            if (diffManifestFile.exists()) {
+                String currentPackageFolderPath = getCurrentPackageFolderPath();
+                CodePushUtils.mergeEntriesInFolder(currentPackageFolderPath, newPackageFolderPath);
+                WritableMap diffManifest = CodePushUtils.getWritableMapFromFile(diffManifestFilePath);
+                ReadableArray deletedFiles = diffManifest.getArray("deletedFiles");
+                for (int i = 0; i < deletedFiles.size(); i++) {
+                    String fileNameToDelete = deletedFiles.getString(i);
+                    File fileToDelete = new File(newPackageFolderPath, fileNameToDelete);
+                    CodePushUtils.deleteFileSilently(fileToDelete);
+                }
+            }
+
+            CodePushUtils.mergeEntriesInFolder(unzippedFolderPath, newPackageFolderPath);
+            CodePushUtils.deleteFileAtPathSilently(unzippedFolderPath);
+            String relativeBundlePath = findMainBundleInFolder(newPackageFolderPath);
+
+            if (relativeBundlePath == null) {
+                throw new CodePushInvalidPackageException();
+            } else {
+                JSONObject updatePackageJSON = CodePushUtils.convertReadableToJsonObject(updatePackage);
+                try {
+                    updatePackageJSON.put(RELATIVE_BUNDLE_PATH_KEY, relativeBundlePath);
+                } catch (JSONException e) {
+                    throw new CodePushUnknownException("Unable to set key " +
+                            RELATIVE_BUNDLE_PATH_KEY + " to value " + relativeBundlePath +
+                            " in update package.", e);
+                }
+                updatePackage = CodePushUtils.convertJsonObjectToWriteable(updatePackageJSON);
+            }
+        } else {
+            // File is not a zip.
+            File updateBundleFile = new File(newPackageFolderPath, UPDATE_BUNDLE_FILE_NAME);
+            downloadFile.renameTo(updateBundleFile);
+        }
+
+        String bundlePath = CodePushUtils.appendPathComponent(newPackageFolderPath, PACKAGE_FILE_NAME);
+        CodePushUtils.writeReadableMapToFile(updatePackage, bundlePath);
+    }
+
+    public String findMainBundleInFolder(String folderPath) {
```

Comment:
How about "findJSBundleInUpdateContents"? The word "Main" doesn't seem necessary and I like adding "UpdateContents" since it adds a little more semantic value to what this function is trying to achieve.